src/engine/sparqlExpressions/AggregateExpression.h

```diff
@@ -23,25 +23,70 @@ inline auto identity = [](auto&& result, size_t) { return AD_FWD(result); };
 
 namespace detail {
 
+template <typename OperandGenerator>
+class UniqueElementsRange;
+
+template <typename OperandGenerator>
+auto createUniqueElementsRange(const EvaluationContext* context,
+                               size_t inputSize,
+                               OperandGenerator operandGenerator) {
+  using ElementType = ql::ranges::range_value_t<OperandGenerator>;
+  return ad_utility::InputRangeTypeErased<ElementType>(
+      UniqueElementsRange<OperandGenerator>(context, inputSize,
+                                            std::move(operandGenerator)));
+}
+
 // For a given `operandGenerator`, generate the sequence of distinct values.
 // This is needed for aggregation together with the `DISTINCT` keyword. For
 // example, `COUNT(DISTINCT ?x)` should count the number of distinct values for
 // `?x`.
 inline auto getUniqueElements = [](const EvaluationContext* context,
-                                   size_t inputSize, auto operandGenerator)
-    -> cppcoro::generator<
-        ql::ranges::range_value_t<decltype(operandGenerator)>> {
-  using OperandGenerator = decltype(operandGenerator);
-  ad_utility::HashSetWithMemoryLimit<
-      ql::ranges::range_value_t<OperandGenerator>>
-      uniqueHashSet(inputSize, context->_allocator);
-  for (auto& operand : operandGenerator) {
-    if (uniqueHashSet.insert(operand).second) {
-      auto elForYielding = std::move(operand);
-      co_yield elForYielding;
```

Comment:
This is a fancy transform + filter, but with stateful functions, so this probably doesn't work out of the box. 
I will think of something.